activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -110,5 +110,23 @@ def self.add_destroy_callbacks(model, reflection)
       name = reflection.name
       model.after_destroy lambda { |o| o.association(name).handle_dependency }
     end
+
+    def self.define_validations(model, reflection)
+      if reflection.options.key?(:required)
+        reflection.options[:optional] = !reflection.options.delete(:required)
+      end
+
+      if reflection.options[:optional].nil?
+        required = model.require_belongs_to
+      else
+        required = !reflection.options[:optional]
+      end
```

Comment:
@jeremy I try to rethink this flow, but deprecation was marked as unnecessary by @dhh here.